### PR TITLE
Jetpack Boost: Improve DataSync error handling

### PR DESCRIPTION
**Pull request by pyronaur**
## Proposed changes:
This lays the groundwork for improved error management in DataSync.

Previously it was really difficult diagnosing where DataSync errors come from because schema parsers had no idea of what's going on and when they're being parsed int he call tree.

In addition to that - it was impossible to communicate to the front end by just throwing errors from DataSync entries.

This introduces two new error classes:

`Schema_Internal_Error` for internal parsing errors - they will be caught within DataSync parsing process and converted into `Schema_Parsing_Error` augmenting the object with a bit of data along the way so that it's possible to trace exactly what went wrong, what was expected and what was passed. 

DataSync schemas are also now JSON Serializable, meaning that if you `json_encode` a schema, you get back a JSON representation of the schema. This could possibly be used for additional debugging utilities like Zod vs Backend schema validator, but that's for a later date.

For right now, here's a brief overview what's changed:

* Added Error classes
* Every type is now serializable and inspectable as a string
* Added a lot of unit tests for error handling
* JSON Serializable Schemas

Here's a list of things that still aren't done, but now that the groundwork is there and stable, I'll be able to come up with follow-up PRs:

* Full Backend to Zod error handling compatibility. Throw errors in DataSync backend, and gracefully receive them in `useDataSync()`
* "Debug Mode" - while this PR contains a single "error_log" that's going to log validation errors with a few more details - this is still not done. We'll want more "dev mode safety", like fallback value validation when WP_DEBUG is enabled (because currently our fallback values are incorrect, like expecting an array but returning null, etc. )
* Proper error handling at the endpoint level. 
* Documentation

Ideas for the future:
JSON Deserializing - if there's enough time left, I wonder if we can replace `Schema::as_somehting()` with a PHP array to also deserialize it. It's actually not impossible at this point to create a CLI tool that reads a Zod schema and creates a PHP array and ensures that they're in sync.

### 🦋 Side-quest
As I was figuring errors out, I accidentally went on a side quest, replacing all `json_decode( $value )` calls with `json_decode( $value, ARRAY_A )`.

Anonymous objects in PHP aren't great:

```
wp> $obj = new stdClass();
=> object(stdClass)#8529 (0) {
}
wp> empty($obj);
=> bool(false)
wp> !!$obj
=> bool(true)
wp> $arr = array();
=> array(0) {
}
wp> empty($arr);
=> bool(true)
wp> !!$arr
=> bool(false)
wp>
```

This can produce bugs in case an object isn't handled properly. 

```
wp> $x = json_decode("{}");
=> object(stdClass)#8531 (0) {
}
wp> empty($x);
=> bool(false)
wp> $y = json_decode("{}", ARRAY_A);
=> array(0) {
}
wp> empty($y);
=> bool(true)
wp>
```

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
1. Apply patch
2. Build everything
3. No really, build everything
4. Test that everything works as expected in Boost Plugin
5. Dump schemas and produce errors - have a look at the logs and enjoy that you finally don't have to hunt for which key/schema has failed.

### CODE REVIEW

File: projects/packages/wp-js-data-sync/src/schema/modifiers/class-decorate-with-default.php [hunk 1]
```diff
@@ -18,15 +18,30 @@ public function __construct( Parser $parser, $default_value ) {
...
 
 	public function get_default_value() {
 		return $this->default_value;
 	}
+
+	public function __toString() {
+		return $this->parser->__toString() . ' (default: ' . $this->default_value . ')';
+	}
+
+	#[\ReturnTypeWillChange]
```
Comment (thingalon) [inline projects/packages/wp-js-data-sync/src/schema/modifiers/class-decorate-with-default.php:41]: https://php.watch/versions/8.1/ReturnTypeWillChange

Does this belong above or in the function? I'm new to these tags.

File: projects/packages/wp-js-data-sync/tests/php/schema/integration/test-integration-parsing-errors.php [hunk 2]
```diff
@@ -0,0 +1,139 @@
+<?php
+
+use Automattic\Jetpack\WP_JS_Data_Sync\Schema\Schema;
+use Automattic\Jetpack\WP_JS_Data_Sync\Schema\Schema_Parsing_Error;
+use Automattic\Jetpack\WP_JS_Data_Sync\Schema\Schema_Validation_Meta;
+use PHPUnit\Framework\TestCase;
+
+class Test_Integration_Parsing_Errors extends TestCase {
```
Comment (thingalon) [inline projects/packages/wp-js-data-sync/tests/php/schema/integration/test-integration-parsing-errors.php:8]: ❤️ tests

```
> ./vendor/phpunit/phpunit/phpunit --colors=always
PHPUnit 9.6.15 by Sebastian Bergmann and contributors.

..........................                                        26 / 26 (100%)

Time: 00:00.004, Memory: 34.00 MB

OK (26 tests, 96 assertions)
```

Warms my heart.

**Review by thingalon — CHANGES_REQUESTED** [review]
Nice work. Very happy to see proper error messages in DataSync :)

Got a bug. I'm getting this from `image-guide/reports/latest`:
```
[17-Jan-2024 04:43:55 UTC] PHP Fatal error:  Uncaught Error: Cannot use object of type stdClass as array in /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/modules/image-size-analysis/data-sync/Image_Size_Analysis_Summary.php:44
Stack trace:
#0 /usr/local/src/jetpack-monorepo/projects/packages/wp-js-data-sync/src/class-data-sync-entry-adapter.php(58): Automattic\Jetpack_Boost\Modules\Image_Size_Analysis\Data_Sync\Image_Size_Analysis_Summary->get()
#1 /usr/local/src/jetpack-monorepo/projects/packages/wp-js-data-sync/src/endpoints/class-endpoint.php(157): Automattic\Jetpack\WP_JS_Data_Sync\Data_Sync_Entry_Adapter->get()
#2 /usr/local/src/jetpack-monorepo/projects/packages/wp-js-data-sync/src/endpoints/class-endpoint.php(104): Automattic\Jetpack\WP_JS_Data_Sync\Endpoints\Endpoint->handler()
#3 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1193): Automattic\Jetpack\WP_JS_Data_Sync\Endpoints\Endpoint->handle_get()
#4 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(1041): WP_REST_Server->respond_to_request()
#5 /var/www/html/wp-includes/rest-api/class-wp-rest-server.php(431): WP_REST_Server->dispatch()
#6 /var/www/html/wp-includes/rest-api.php(424): WP_REST_Server->serve_request()
#7 /var/www/html/wp-includes/class-wp-hook.php(324): rest_api_loaded()
#8 /var/www/html/wp-includes/class-wp-hook.php(348): WP_Hook->apply_filters()
#9 /var/www/html/wp-includes/plugin.php(565): WP_Hook->do_action()
#10 /var/www/html/wp-includes/class-wp.php(418): do_action_ref_array()
#11 /var/www/html/wp-includes/class-wp.php(813): WP->parse_request()
#12 /var/www/html/wp-includes/functions.php(1336): WP->main()
#13 /var/www/html/wp-blog-header.php(16): wp()
#14 /var/www/html/index.php(17): require('...')
#15 {main}
```

I also see a bunch of red X's at the bottom of the PR. Those need to be cleaned up.

File: projects/packages/wp-js-data-sync/src/schema/modifiers/class-decorate-with-default.php (hunk 1 above)
Comment (pyronaur) [reply projects/packages/wp-js-data-sync/src/schema/modifiers/class-decorate-with-default.php:41]: Right above the function definition

**Comment by pyronaur** [comment]
Tracked down a critical css bug, but there are still e2e tests that fail. I'll keep digging tomorrow.

**Comment by pyronaur** [comment]
Yay! Tests pass - ready to merge.

**Review by thingalon — APPROVED** [review]
Lookin' good, and you were right that the updates to the companion plugin fixed the issues I ran into. Let's roll!

**Pull request merged**